service/storage_service.cc

```diff
@@ -6153,6 +6189,165 @@ future<> storage_service::cleanup_tablet(locator::global_tablet_id tablet) {
     });
 }
 
+future<join_node_request_result> storage_service::join_node_request_handler(join_node_request_params params) {
+    join_node_request_result result;
+    slogger.info("raft topology: received request to join from host_id: {}", params.host_id);
+
+    if (params.cluster_name != _db.local().get_config().cluster_name()) {
+        result.result = join_node_request_result::rejected{
+            .reason = ::format("Cluster name check failed. This node cannot join the cluster "
+                                "because it expected cluster name \"{}\" and not \"{}\"",
+                                params.cluster_name,
+                                _db.local().get_config().cluster_name()),
+        };
+        co_return result;
+    }
+
+    if (params.snitch_name != _db.local().get_snitch_name()) {
+        result.result = join_node_request_result::rejected{
+            .reason = ::format("Snitch name check failed. This node cannot join the cluster "
+                                "because it uses \"{}\" and not \"{}\"",
+                                params.snitch_name,
+                                _db.local().get_snitch_name()),
+        };
+        co_return result;
+    }
+
+    co_await _topology_state_machine.event.when([this] {
+        // The first node defines the cluster and inserts its entry to the
+        // `system.topology` without checking anything. It is unlikely but
+        // possible that the `join_node_request_handler` fires before the first
+        // node inserts its entry, therefore we might need to wait
+        // until that happens, here.
+        return !_topology_state_machine._topology.is_empty();
+    });
+
+    while (true) {
+        std::optional<group0_guard> guard;
+        try {
+            guard.emplace(co_await _group0->client().start_operation(&_abort_source));
+        } catch (abort_requested_exception&) {
+            throw;
+        } catch (...) {
+            slogger.warn("raft topology: join_node_request: failed to start group0 operation: {}, retrying.",
```

Comment:
I had the following failure in `test_replace.not_sure_which_test_case`:

- Cluster with nodes `A`, `B`, `C` - `A` is the topology coordinator/group0 leader,
- `A` is killed,
- New node `D` attempts to replace `A` with the same IP as `A`, sends `join_node_request` rpc to node `B`,
- Node `B` handles the RPC and wants to perform group0 operation and wants to perform a barrier - still thinks that `A` is the leader and is alive, sends an RPC to its IP,
- `D` accidentally receives the request that was meant to `A` but throws an exception because of host_id mismatch,
- Failure is propagated back to `B`, and then to `A` - and `A` fails the replace operation.

Adding the retries makes the logic resilient to this. If you have another idea how to avoid this error, I'm all ears (eyes?).

Reply:
Wow, interesting sequence of events.

I think retrying is OK, but unlike `group0_concurrent_modification`, which always implies progress, I don't think we should retry `start_operation` failure forever. Maybe a timeout is appropriate here.

Also what was the type of exception that you got in this scenario? Maybe we should only catch that one instead of `...`.

Reply:
```
 WARN  2023-09-22 08:28:37,920 [shard 0:goss] storage_service - raft topology: join_node_request: failed to start group0 operation: seastar::rpc::remote_verb_error (Got message for server 39e64975-cde9-4c17-9601-d82913046ddf, but my id is 5b46ceba-346d-4cf0-98b1-da424f268d23), retrying.
 ```

It's `seastar::rpc::remote_verb_error`.

> I think retrying is OK, but unlike `group0_concurrent_modification`, which always implies progress, I don't think we should retry `start_operation` failure forever. Maybe a timeout is appropriate here.

I can add a timeout, though I wonder if there is a better way.

Reply:
@gleb-cloudius maybe you have some better idea?

Reply:
Scenario is indeed weird. Worth keeping in mind. Ideas? Well here is one: `join_node_request` checks if the id of the node being replaced is currently considered to be a leader wait Nsecs for leader change. Otherwise return an error.

Reply:
If ID of the leader == ID of the replaced node, then it's either our case or an operator error (trying to replace a node that is alive). In the latter case, we will delay the error by N seconds (according to Kamil node reelection usually takes ~ 3s, so N >> 3?) - which is not terrible, though maybe can be avoided.

Instead of looking at IDs, maybe we could compare IPs? If the IP of the _new_ node is the same as IP of the perceived raft leader, then we could wait for a leader change.

Reply:
We want to get rid of id to ip translation in upper layer. I think waiting a couple of seconds before failing is fine.

Reply:
Ack, I'm now comparing IDs and waiting until the leader changes.